src/adapter/philips_hue/api/mod.rs

```diff
@@ -0,0 +1,234 @@
+        // TODO // ___THIS IS A GAPING SECURITY HOLE!___ // TODO //
+        // TODO //                                       // TODO //
+
+        // TODO: replace with a loooong stored random string
+        let stored_random = String::from("4"); // chosen by fair dice roll.
+        let token_gen = HueApiToken { iv: stored_random, id: id.clone() };
+        let token = format!("foxbox-{}", token_gen.as_hash());
+        HueApi { id: id, ip: ip, token: token }
+    }
+
+    pub fn get(&self, cmd: String) -> Result<String, Box<Error>> {
+        let uri = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("GET request to Philips Hue bridge {}: {}", self.id, uri);
+        let content = http::get(uri);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn post(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let uri = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("POST request to Philips Hue bridge {}: {} data: {}", self.id, &uri, &data);
+        let content = http::post(uri, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn post_unauth(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let uri = format!("http://{}/{}", self.ip, cmd);
+        debug!("POST request to Philips Hue bridge {}: {} data: {}", self.id, &uri, &data);
+        let content = http::post(uri, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn put(&self, cmd: String, data: String) -> Result<String, Box<Error>> {
+        let uri = format!("http://{}/api/{}/{}", self.ip, self.token, cmd);
+        debug!("PUT request to Philips Hue bridge {}: {} data: {}", self.id, &uri, &data);
+        let content = http::put(uri, data);
+        debug!("Philips Hue API response: {:?}", content);
+        content
+    }
+
+    pub fn is_available(&self) -> bool {
+        let uri = format!("http://{}/", self.ip);
+        let content = http::get(uri);
+        match content {
+            Ok(val) => {
+                val.contains("hue personal wireless lighting")
+            },
+            Err(_) => {
+                false
+            }
+        }
+    }
+
+    pub fn get_settings(&self) -> String {
+        // [{"error":{"type":1,"address":"/","description":"unauthorized user"}}]
+        self.get("".to_owned()).unwrap_or("".to_owned()) // TODO no unwrap
+    }
+
+    pub fn is_paired(&self) -> bool {
+        let s = self.get_settings();
+        !s.contains("unauthorized user")
+    }
+
+    pub fn pair(&self) -> String {
+        // [{"success":{"username":"foxboxb-001788fffe25681a"}}]
+        // [{"error":{"type":101,"address":"/","description":"link button not pressed"}}]
+        let url = format!("api");
+        let req = format!("{{\"username\": \"{}\", \"devicetype\": \"foxbox_hub\"}}", self.token);
```

Comment:
that's fragile in general because you don't escape potential special characters from the token.